File: T-Bank/Algorithms-and-Data-Structures-2025/4sprint-SolvingProblemMethodsPart1/contest/E-Разделение_массива/solution.cpp
#include <iostream>
#include <vector>


bool canDistribute(int& n, int k, std::vector<long long>& nums, long long& bound) {
    --k;
    long long curSum = 0;
    for(long long& num : nums) {
        if (curSum + num > bound) {
            curSum = num;
            --k;
        } else curSum += num;
    }
    return k >= 0;
}


void solution() {
    /*
    Time Complexity O(NlogM)
    Memory Complexity O(1)
    */
    int n, k;
    std::scanf("%d %d", &n, &k);
    std::vector<long long> nums (n, 0);
    long long left = 0;
    long long right = 0;
    for (int i = 0; i < n; ++i) {
        std::scanf("%lld", &nums[i]);
        right += nums[i];
        left = std::max(left, nums[i]);
    }
    while (left <= right) {
        long long middle = left + (right - left) / 2;
        if (canDistribute(n, k, nums, middle)) {
            right = middle - 1;
        } else left = middle + 1;
    }
    std::printf("%lld\n", left);
}


int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/4sprint-SolvingProblemMethodsPart1/contest/E-Разделение_массива/solution.py
import sys


def can_distribute(n: int, k: int, nums: list[int], bound: int) -> bool:
    k -= 1
    cur_sum: int = 0
    for num in nums:
        if cur_sum + num > bound:
            k -= 1
            cur_sum = num
        else: cur_sum += num
    return k >= 0


def solution() -> None:
    """
    Time Complexity O(NlogM)
    Memory Complexity O(1)
    """
    n, k = map(int, sys.stdin.readline().rstrip().split())
    nums: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    left: int = max(nums)
    right: int = sum(nums)
    while left <= right:
        middle: int = left + ((right - left) >> 1)
        if can_distribute(n, k, nums, middle): right = middle - 1
        else: left = middle + 1
    sys.stdout.write('{}\n'.format(left))


if __name__ == '__main__':
    solution()